### PR TITLE
Corrosion

### PR DESCRIPTION
**Pull request by AsparagusEduardo**

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
It's only missing its Toxic Orb/Fling and Magic Coat/Bounce interaction, but everything else is cool.

**Comment by DizzyEggg** [comment]
I think a macro/function would be better. For both scripting and code

**Comment by AsparagusEduardo** [comment]
How about now?

File: asm/macros/battle_script.inc [hunk 1]
```diff
@@ -1713,6 +1713,12 @@
 	various \battler, VARIOUS_JUMP_IF_SHIELDS_DOWN_PROTECTED
 	.4byte \ptr
 	.endm
+
+	.macro handletypeimmunity battlerability:req, battlertypes:req, status:req
```
Comment (DizzyEggg) [inline asm/macros/battle_script.inc:1717]: I think `trypoisontype` would be a better name. Also would prefer if the battlescript pointer would be passed as an argument.
Would also get rid of the status argument and treat it only for trying to inflict Poison/Bad Poison

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -6667,6 +6665,23 @@ static void HandleTerrainMove(u32 moveEffect)
     }
 }
 
+bool32 DoesAbilityBypassTypeImmunity(u8 battlerAttacker, u8 battlerTarget, u8 status)
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:6668]: Why not just `CanPoisonType`? Without the status argument

File: src/battle_script_commands.c [hunk 3]
```diff
@@ -6667,6 +6665,23 @@ static void HandleTerrainMove(u32 moveEffect)
...
+{
+    switch(status)
+    {
+        case STATUS1_POISON:
+        case STATUS1_TOXIC_POISON:
+            if (GetBattlerAbility(battlerAttacker) == ABILITY_CORROSION)
+                return TRUE;
+            else if (IS_BATTLER_OF_TYPE(battlerTarget, TYPE_POISON))
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:6676]: Could be simplified to
return (GetBattlerAbility(battlerAttacker) == ABILITY_CORROSION
|| !(IS_BATTLER_OF_TYPE(battlerTarget, TYPE_POISON) && IS_BATTLER_OF_TYPE(battlerTarget, TYPE_STEEL)))

File: asm/macros/battle_script.inc (hunk 1 above)
Comment (AsparagusEduardo) [reply asm/macros/battle_script.inc:1717]: My mindset was making it so it could support custom abilities similar to corrosion, but sure.

Comment (AsparagusEduardo) [reply asm/macros/battle_script.inc]: Done.

File: src/battle_script_commands.c (hunk 3 above)
Comment (AsparagusEduardo) [reply src/battle_script_commands.c]: Done, but changed the final && to ||.

File: src/battle_script_commands.c (hunk 2 above)
Comment (AsparagusEduardo) [reply src/battle_script_commands.c]: Done.

File: asm/macros/battle_script.inc [hunk 4]
```diff
@@ -1713,6 +1713,12 @@
 	various \battler, VARIOUS_JUMP_IF_SHIELDS_DOWN_PROTECTED
 	.4byte \ptr
 	.endm
+
+	.macro trypoisontype battlerability:req, battlertypes:req, ptr:req
```
Comment (DizzyEggg) [inline asm/macros/battle_script.inc:1717]: Could battlerability battlertypes be renamed to simply attacker and target?

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -2254,7 +2254,7 @@ void SetMoveEffect(bool32 primary, u32 certain)
                 }
                 RESET_RETURN
             }
-            if ((IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_POISON) || IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_STEEL))
+            if (CanPoisonType(gBattleScripting.battler, gEffectBattler)
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:2257]: !CanPoisonType

File: src/battle_script_commands.c [hunk 6]
```diff
@@ -2264,9 +2264,7 @@ void SetMoveEffect(bool32 primary, u32 certain)
                 gBattleCommunication[MULTISTRING_CHOOSER] = 2;
                 RESET_RETURN
             }
-            if (IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_POISON))
-                break;
-            if (IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_STEEL))
+            if (CanPoisonType(gBattleScripting.battler, gEffectBattler))
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:2267]: !CanPoisonType

File: src/battle_script_commands.c [hunk 7]
```diff
@@ -2401,7 +2399,7 @@ void SetMoveEffect(bool32 primary, u32 certain)
                 }
                 RESET_RETURN
             }
-            if ((IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_POISON) || IS_BATTLER_OF_TYPE(gEffectBattler, TYPE_STEEL))
+            if (CanPoisonType(gBattleScripting.battler, gEffectBattler)
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:2402]: !CanPoisonType

File: src/battle_script_commands.c [hunk 8]
```diff
@@ -6885,6 +6890,14 @@ static void Cmd_various(void)
         gBattleStruct->friskedBattler = 0;
         gBattleStruct->friskedAbility = FALSE;
         break;
+    case VARIOUS_HANDLE_TYPE_IMMUNITY:
+        gBattlerTarget = GetBattlerForBattleScript(gBattlescriptCurrInstr[3]);
+
+        if (!CanPoisonType(gActiveBattler, gBattlerTarget))
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:6896]: Could just do

if (!CanPoisonType(gActiveBattler, GetBattlerForBattleScript(gBattlescriptCurrInstr[3])))

File: src/battle_util.c [hunk 9]
```diff
@@ -4865,7 +4865,7 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
         {
         case HOLD_EFFECT_TOXIC_ORB:
             if (!gBattleMons[battlerId].status1
-                && !IS_BATTLER_OF_TYPE(battlerId, TYPE_POISON) && !IS_BATTLER_OF_TYPE(battlerId, TYPE_STEEL)
+                && CanPoisonType(battlerId, gEffectBattler)
```
Comment (DizzyEggg) [inline src/battle_util.c:4868]: I don't actually think Corrosion applies to Toxic Orb, so I would just leave the type checks theere

Comment (KDSKardabox) [reply src/battle_util.c:4868]: It does according to Bulbapedia but haven't tested it in-game.

File: asm/macros/battle_script.inc (hunk 4 above)
Comment (AsparagusEduardo) [reply asm/macros/battle_script.inc:1717]: Done.

File: src/battle_script_commands.c (hunk 5 above)
Comment (AsparagusEduardo) [reply src/battle_script_commands.c:2257]: Done.

File: src/battle_script_commands.c (hunk 6 above)
Comment (AsparagusEduardo) [reply src/battle_script_commands.c:2267]: Done.

File: src/battle_script_commands.c (hunk 7 above)
Comment (AsparagusEduardo) [reply src/battle_script_commands.c:2402]: Done.

File: src/battle_util.c (hunk 9 above)
Comment (AsparagusEduardo) [reply src/battle_util.c:4868]: Found this video of somebody testing this in SM and indeed Toxic Orb can poison if the holder has Corrosion.
https://www.youtube.com/watch?v=IjFHFBXUpV8
Interestingly, they also tested Magic Bounce's effect and the bounced-back attack can't poison even if it was originally by a Corrosion owner

Comment (DizzyEggg) [reply src/battle_util.c:4868]: Alright then, shouldn't it be CanPoisonType(battlerId, battlerId) then?

Comment (AsparagusEduardo) [reply src/battle_util.c:4868]: Tested and worked both ways. Do I just leave it as (battlerId, battlerId)?

**Pull request merged**